### DISCUSSION TITLE
Which URL should I send my request to? Which ID and key are needed?

### DISCUSSION THREAD

**Discussion by Lollichkeit** [amzn/ads-advanced-tools-docs#151] [Q&A]
Hello,

I need help.

I'm already logged in to Amazon Ads and Amazon Developer.
My Amazon Ads API has been applied for and approved.
I created a security profile at Amazon Developer.

But now I can't go any further.

I don't know what data I need to send a request with PHP and then get data back.

If anyone could help me with this I would be very grateful.

**Comment by jerirose** [comment]
Hi @Lollichkeit , first you'll need to make sure that you've [assigned API access to your Login With Amazon application as described here](https://advertising.amazon.com/API/docs/en-us/guides/onboarding/assign-api-access). (You may have already done this.)

Once this is complete, you can use the client ID and client secret of your Login With Amazon application to generate an authorization grant from an Amazon Ads account, then use that grant to retrieve an access token. Your request for an access token will be sent to [the Login With Amazon server for your region as described here](https://advertising.amazon.com/API/docs/en-us/guides/get-started/retrieve-access-token).

Using your client ID and an access token, you can make calls to the Amazon Ads API -- these requests will be directed to the [regional host for the Amazon Ads account](https://advertising.amazon.com/API/docs/en-us/reference/api-overview#typical-use-cases) whose data you want to retrieve. 

These first steps are described in the [**Get started** section of the API documentation](https://advertising.amazon.com/API/docs/en-us/guides/get-started/overview). You may also find it useful to try the [Postman quickstart guide](https://advertising.amazon.com/API/docs/en-us/guides/get-started/using-postman-collection) to test your first calls.

You'll use this same pattern of credentials to call the API programmatically using PHP or any other interface or language. Please feel free to reply in this thread if you run into other specific issues!

**Comment by Lollichkeit** [comment]
I think I understood it. Thank you very much, I just don't know what I have to enter in the "Allowed Return URLs".

**Comment by jerirose** [comment]
Hi @Lollichkeit , an [allowed return URL](https://advertising.amazon.com/API/docs/en-us/guides/get-started/create-authorization-grant#what-is-an-allowed-return-url) can be _any_ `https` URL -- but the `redirect_uri` specified in your authorization grant request must match one of the **Allowed Return URLs** you have set for the security profile in Login with Amazon.

As an example, the user in the videos shown in the [Postman quickstart guide](https://advertising.amazon.com/API/docs/en-us/guides/get-started/using-postman-collection) simply uses `https://amazon.com`, so the browser is redirected to the Amazon homepage, and the authorization grant code is manually retrieved from the resulting URL.

For robust applications that are designed to interact with multiple Amazon Ads customers, the return URL might instead lead to an application page that programmatically handles the grant code. But for simpler use cases, it is likely easiest to manually retrieve the code as described above.

**Reply by Lollichkeit** [reply]
I tried to set up the postman now, but then when refreshing token the problem I have this error in the url:
`?error_description=An+unknown+scope+was+requested&state=State&error=invalid_scope`

I'm wondering now but can't you somehow get this automatically with PHP the code directly. I just want to get some values from the advertising of my company.

**Reply by jerirose** [reply]
Hi @Lollichkeit , you may find some answers to the "unknown scope" issue in this discussion: https://github.com/amzn/ads-advanced-tools-docs/discussions/161

In that particular case, the API caller had missed the [assign API access to your LwA app](https://advertising.amazon.com/API/docs/en-us/guides/onboarding/assign-api-access) step, in which case Login With Amazon will not allow the client ID to request access to the `advertising::campaign_management` scope.

**Reply by Lollichkeit** [reply]
So if I understand correctly, I need to connect the API to the LWA via the email used to activate the API, but I only get to the following web page where I can't connect my API to the LWA.

![image](https://github.com/amzn/ads-advanced-tools-docs/assets/64012478/0c7b04b8-c428-4dd6-9645-ae6822cba50f)

How to connect the API with LWA?

**Reply by jerirose** [reply]
@Lollichkeit This indicates that your API access has been successfully applied to an LwA client application. Does the client ID that you see on this screen match the client ID you are using to connect to the API? The client ID you provide to the Postman environment should be the same one that you see on this **Congratulations!** screen.

Can you post your entire authorization grant URL here? (You can put `XXXXX` in place of your client ID if you prefer.)

**Reply by Lollichkeit** [reply]
The LWA client ID and this one from Amazon ADs are two different ones. I did not receive a client secret for the Amazon ADs ID either.
What can I do now?

**Reply by jerirose** [reply]
@Lollichkeit There are a few possibilities here that relate to the [assign API access](https://advertising.amazon.com/API/docs/en-us/guides/onboarding/assign-api-access) step:
1) Someone else in your organization may have opened the access approval email from Amazon and assigned access to a different LwA application.
2) You may have more than one LwA security profile in your own Amazon Developer account.

If one of those is the case and you are able to get the client secret for the LwA client ID shown in your screenshot above, you should be able to then generate an authorization grant code for the appropriate scope using that client ID.

If you're not able to find the information for that client ID, we recommend submitting a ticket to the [Amazon Ads API Jira service desk](https://amzn-clicks.atlassian.net/servicedesk/customer/portal/2/group/2). They will be able to guide you through the next steps.

**Comment by Lollichkeit** [comment]
First of all, thank you for your help.
I still don't understand the topic. Can you help me with the refresh token first? Maybe then I'll understand.

My code:
```
$client = new \GuzzleHttp\Client();
$response = $client->request('POST', 'https://api.amazon.co.uk/auth/o2/token?grant_type=??????????????&code=???????????????&redirect_uri=https://amazon.com&client_id='.$lwaID.'&client_secret='.$lwaSecret );
```
I would first like to receive the token with Guzzle. I need that to get further information, right?

**Reply by jerirose** [reply]
The user interface steps described in the onboarding documentation are necessary for the _initial_ authorization grant. However, after you have retrieved an access token and refresh token using an auth grant code _once_, you can use the refresh token to programmatically retrieve new access tokens in your PHP application.

You can find information about using refresh tokens in the [**Access tokens** document](https://advertising.amazon.com/API/docs/en-us/guides/account-management/authorization/access-tokens#generating-an-access-token-using-a-refresh-token) in the advanced tools center. That section includes an example curl request for using a refresh token, which you can adapt to use with Guzzle by including the required values in your URL string. Assuming you have the refresh token value stored as the variable `$refreshToken`, your code might look something like this:

```php
$client = new \GuzzleHttp\Client();
$response = $client->request('POST', 'https://api.amazon.co.uk/auth/o2/token?grant_type=refresh_token&refresh_token=' . $refreshToken . '&redirect_uri=https://amazon.com&client_id=' . $lwaID . '&client_secret=' . $lwaSecret );
```
